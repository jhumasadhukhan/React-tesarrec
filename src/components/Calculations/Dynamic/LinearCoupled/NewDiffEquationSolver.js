//DyByDx= x+3y

const NewDiffEquationSolver = (props) => {
  /**
   * A method component that takes in a equation and outputs a chart
   */

  var Integrators = {
    Euler: [[1]],
    Midpoint: [
      [0.5, 0.5],
      [0, 1],
    ],
    Heun: [
      [1, 1],
      [0.5, 0.5],
    ],
    Ralston: [
      [2 / 3, 2 / 3],
      [0.25, 0.75],
    ],
    K3: [
      [0.5, 0.5],
      [1, -1, 2],
      [1 / 6, 2 / 3, 1 / 6],
    ],
    SSP33: [
      [1, 1],
      [0.5, 0.25, 0.25],
      [1 / 6, 1 / 6, 2 / 3],
    ],
    SSP43: [
      [0.5, 0.5],
      [1, 0.5, 0.5],
      [0.5, 1 / 6, 1 / 6, 1 / 6],
      [1 / 6, 1 / 6, 1 / 6, 1 / 2],
    ],
    RK4: [
      [0.5, 0.5],
      [0.5, 0, 0.5],
      [1, 0, 0, 1],
      [1 / 6, 1 / 3, 1 / 3, 1 / 6],
    ],
    RK38: [
      [1 / 3, 1 / 3],
      [2 / 3, -1 / 3, 1],
      [1, 1, -1, 1],
      [1 / 8, 3 / 8, 3 / 8, 1 / 8],
    ],
    RKF: [
      [1 / 4, 1 / 4],
      [3 / 8, 3 / 32, 9 / 32],
      [12 / 13, 1932 / 2197, -7200 / 2197, 7296 / 2197],
      [1, 439 / 216, -8, 3680 / 513, -845 / 4104],
      [1 / 2, -8 / 27, 2, -3544 / 2565, 1859 / 4104, -11 / 40],
      [25 / 216, 0, 1408 / 2565, 2197 / 4104, -1 / 5, 0],
    ],
  };

  const func = (t, y) => {
    // for a given t , list of dependent varibales

    let eqnResultsArr = [];
    let coordinate = {}; //t initial value
    //generates the eval object
    for (let i = 0; i < y.length; i++) {
      // const dependentVariable = props.modelObj.Config.lineNames[i]; // { a, b , c}
      const dependentVariable = lineNames[i]; // { a, b , c}

      coordinate[dependentVariable] = y[i];
    }

    coordinate[independentLatexForm] = t; // { a, b , c, t}
    const accumulative = {
      ...coordinate,
      // ...props.vars,
      ...constants,
      // t:t
    };

    for (let idx = 0; idx < props.modelObj.Eqns.length; idx++) {
      const eqnObj = props.modelObj.Eqns[idx];

      eqnResultsArr.push(eqnObj.parsedEqn.evaluate(accumulative)); // { a :1 , b: 3.3,t:3}
    }

    return eqnResultsArr;
  };

  //let ki be the set of eqns entered by the user
  //let n be the location
  //let i varies upto the order or the number of points within the step n and n+1
  //Compute ki by substituting the value of tn+bi*h and yn+ summation of aij*h*kj (j varies from 1 to i-1) in the set of eqns entered by the user
  //Compute yn+1 = yn + h * summation of wi*ki (i varies from 1 to the order)
  //Return yn+1 and tn+1 for the plot

  const integrate = (m, f, y, t, h) => {
    for (var k = [], ki = 0; ki < m.length; ki++) {
      var _y = y.slice(),
        dt = ki ? m[ki - 1][0] * h : 0;
      for (var l = 0; l < _y.length; l++)
        for (var j = 1; j <= ki; j++)
          _y[l] = _y[l] + h * m[ki - 1][j] * k[ki - 1][l];
      k[ki] = f(t + dt, _y, dt);
    }
    // eslint-disable-next-line
    for (var r = y.slice(), l = 0; l < _y.length; l++)
      for (var j = 0; j < k.length; j++)
        r[l] = r[l] + h * k[j][l] * m[ki - 1][j];
    return r;
  };

  // const integrate = (meth, func, y, t, h) => {
  //   for (var k = [], ki = 0; ki < meth.length; ki++) {
  //     // for each technique iterate over the solver coeff
  //     var _y = y.slice();

  //     var dt = ki ? meth[ki - 1][0] * h : 0; // dt=0 when ki=0

  //     for (var l = 0; l < _y.length; l++) {
  //       for (var j = 1; j <= ki; j++) {
  //         _y[l] = _y[l] + h * meth[ki - 1][j] * k[ki - 1][l];
  //       }
  //     }
  //     k[ki] = func(t + dt, _y);
  //   }
  //   console.log(k)
  //   for (
  //     var r = y.slice(), l = 0;
  //     l < _y.length;
  //     l++ // 0-2
  //   )
  //     for (
  //       var j = 0;
  //       j < k.length;
  //       j++ //0-4 (rk4)
  //     )
  //       r[l] = r[l] + h * k[j][l] * meth[ki - 1][j]; //r[0]=r[0]+ 0.5*0.5 *16  //r[0]=r[0]+ 0.5*0.5 *16

  //   return r;
  // };

  // let t0 = parseFloat(props.modelObj.Config.t0);
  let t0 = null;
  let initialConditions = [];
  let constants = {};
  let independentLatexForm=null
  let lineNames=[]
  for (let i = 0; i < props.modelObj.Vars.length; i++) {
    if (props.modelObj.Vars[i].VarType === "Independent") {
      t0 = parseFloat(props.modelObj.Vars[i].VarCurrent);
      independentLatexForm=props.modelObj.Vars[i].LatexForm
    }
    if (props.modelObj.Vars[i].VarType === "Dependent") {
      initialConditions.push(parseFloat(props.modelObj.Vars[i].VarCurrent));
      lineNames.push(props.modelObj.Vars[i].LatexForm)
    }
    if (props.modelObj.Vars[i].VarType === "Constant") {
      constants[props.modelObj.Vars[i].LatexForm] = parseFloat(
        props.modelObj.Vars[i].VarCurrent
      );
    }
  }

  // let returnedY = props.modelObj.Config.initialConditions.slice();
  let returnedY = initialConditions.slice();

  let _returnedY = returnedY.slice();
  _returnedY.push(t0);
  let allY = [_returnedY];

  for (let i = 0; i < props.modelObj.Config.numOfCycles; i++) {
    returnedY = integrate(
      Integrators[props.modelObj.Config.method],
      func,
      returnedY,
      t0,
      parseFloat(props.modelObj.Config.h)
    );
    t0 += parseFloat(props.modelObj.Config.h); // constant step size

    allY.push([...returnedY, t0]);
  }

  return allY;
  //   // const f=(h,y,)=>{

  //   // }
  //   ([0, 0.6249999999999999],
  //   [0.5, 0.9999999999999999],
  //   [1, 1.6249999999999998],
  //   [1.5, 2.4999999999999996],
  //   [2, 3.6249999999999996],
  //   [2.5, 4.999999999999999],
  //   [3, 6.625],
  //   [3.5, 8.500000000000002],
  //   [4, 10.625000000000004],
  //   [4.5, 13.000000000000002],
  //   [5, 15.625000000000002],
  //   [5.5, 18.5],
  //   [6, 21.625000000000004],
  //   [6.5, 25.000000000000004],
  //   [7, 28.625],
  //   [7.5, 32.5],
  //   [8, 36.625],
  //   [8.5, 41.00000000000001],
  //   [9, 45.625],
  //   [9.5, 50.5],
  //   [10, 55.62500000000001],
  //   [10.5, 61],
  //   [11, 66.62499999999999],
  //   [11.5, 72.49999999999997],
  //   [12, 78.62499999999999],
  //   [12.5, 84.99999999999999],
  //   [13, 91.62499999999997],
  //   [13.5, 98.49999999999999],
  //   [14, 105.62499999999999],
  //   [14.5, 112.99999999999997])
  // ];

  // [
  //   [
  //     [0.5, 0.5],
  //     [0.875, 1],
  //     [1.5, 1.5],
  //     [2.375, 2],
  //     [3.5, 2.5],
  //     [4.875, 3],
  //     [6.5, 3.5],
  //     [8.375, 4],
  //     [10.5, 4.5],
  //     [12.875, 5],
  //     [15.5, 5.5],
  //     [18.375, 6],
  //     [21.5, 6.5],
  //     [24.875, 7],
  //     [28.5, 7.5],
  //     [32.375, 8],
  //     [36.5, 8.5],
  //     [40.875, 9],
  //     [45.5, 9.5],
  //     [50.375, 10],
  //     [55.5, 10.5],
  //     [60.875, 11],
  //     [66.5, 11.5],
  //     [72.375, 12],
  //     [78.5, 12.5],
  //     [84.875, 13],
  //     [91.5, 13.5],
  //     [98.375, 14],
  //     [105.5, 14.5],
  //     [112.875, 15],
  //     [120.5, 15.5],
  //   ],
  // ];
};

export default NewDiffEquationSolver;
