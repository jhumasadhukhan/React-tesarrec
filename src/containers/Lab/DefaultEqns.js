const DEFAULTEQNS = [
    {
      id: "qwert",
      line: "a",
      DByDLatex: "\\frac{da}{dt}=",
      LatexEqn: "-\\frac{{K_1}ab}{{K_2}+a}-\\frac{{K_3}}ac{{K_4}+a}",
      TextEqn: "-(K_1*a*b)/(K_2+a)-(K_3)/(a)*c*K_4+a",
      errorMessage: null,
    },
    {
      id: "yuiop",
      line: "b",

      DByDLatex: "\\frac{db}{dt}=",
      LatexEqn: "\\frac{{K_1}ab}{{K_3}+a}-{K_4}b",
      TextEqn: "(K_1*a*b)/(K_3+a)-K_4*b",
      errorMessage: null,
    },
    {
      id: "asdfg",
      line: "c",

      DByDLatex: "\\frac{dc}{dt}=",
      LatexEqn: "\\frac{{K_2}}ac{{K_1}+a}-{K_4}c",
      TextEqn: "(K_2*a*c)/(K_1+a)-K_4*c",
      errorMessage: null,
    },
    {
      id: "hjklz",
      line: "d",

      DByDLatex: "\\frac{dd}{dt}=",
      LatexEqn: "-\\frac{{K_1}}ab{{K_4}+a}-\\frac{{K_3}ac}{{K_2}+a}",
      TextEqn: "-(K_1*a*b)/(K_4+a)-(K_3*a*c)/(K_2+a)",
      errorMessage: null,
    },
  ];

  export default DEFAULTEQNS