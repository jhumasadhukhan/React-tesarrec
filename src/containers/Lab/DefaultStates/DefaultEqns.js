const DEFAULTEQNS = [
    {
      id: "qwert",
      line: "a",
      DByDLatex: "\\frac{da}{dt}=",
      LatexEqn: "-\\frac{{K_1}ab}{{K_2}+a}-\\frac{{K_3}}ac{{K_4}+a}",
      TextEqn: "-(K_1*a*b)/(K_2+a)-(K_3)/(a)*c*K_4+a",
      errorMessage: null,
    },
    {
      id: "yuiop",
      line: "b",

      DByDLatex: "\\frac{db}{dt}=",
      LatexEqn: "\\frac{{K_1}ab}{{K_3}+a}-{K_4}b",
      TextEqn: "(K_1*a*b)/(K_3+a)-K_4*b",
      errorMessage: null,
    },
    {
      id: "asdfg",
      line: "c",

      DByDLatex: "\\frac{dc}{dt}=",
      LatexEqn: "\\frac{{K_2}}ac{{K_1}+a}-{K_4}c",
      TextEqn: "(K_2*a*c)/(K_1+a)-K_4*c",
      errorMessage: null,
    },
    {
      id: "hjklz",
      line: "d",

      DByDLatex: "\\frac{dd}{dt}=",
      LatexEqn: "-\\frac{{K_1}}ab{{K_4}+a}-\\frac{{K_3}ac}{{K_2}+a}",
      TextEqn: "-(K_1*a*b)/(K_4+a)-(K_3*a*c)/(K_2+a)",
      errorMessage: null,
    },
    {
      id: "dfsf",
      line: "f",

      DByDLatex: "\\frac{df}{dt}=",
      LatexEqn: "-\\frac{{K_1}}ab{{K_4}+a}-\\frac{{K_3}ac}{{K_2}+a}",
      TextEqn: "-(K_1*a*b)/(K_4+a)-(K_3*a*c)/(K_2+a)",
      errorMessage: null,
    },
    {
      id: "klsdf",
      line: "g",

      DByDLatex: "\\frac{dg}{dt}=",
      LatexEqn: "-\\frac{{K_1}}ab{{K_4}+a}-\\frac{{K_3}ac}{{K_2}+a}",
      TextEqn: "-(K_1*a*b)/(K_4+a)-(K_3*a*c)/(K_2+a)",
      errorMessage: null,
    },
    {
      id: "sdgfdg",
      line: "h",

      DByDLatex: "\\frac{dh}{dt}=",
      LatexEqn: "-\\frac{{K_1}}ab{{K_4}+a}-\\frac{{K_3}ac}{{K_2}+a}",
      TextEqn: "-(K_1*a*b)/(K_4+a)-(K_3*a*c)/(K_2+a)",
      errorMessage: null,
    },
    {
      id: "ewrtew",
      line: "j",

      DByDLatex: "\\frac{dj}{dt}=",
      LatexEqn: "-\\frac{{K_1}}ab{{K_4}+a}-\\frac{{K_3}ac}{{K_2}+a}",
      TextEqn: "-(K_1*a*b)/(K_4+a)-(K_3*a*c)/(K_2+a)",
      errorMessage: null,
    },
    {
      id: "sjmew",
      line: "k",

      DByDLatex: "\\frac{dk}{dt}=",
      LatexEqn: "-\\frac{{K_1}}ab{{K_4}+a}-\\frac{{K_3}ac}{{K_2}+a}",
      TextEqn: "-(K_1*a*b)/(K_4+a)-(K_3*a*c)/(K_2+a)",
      errorMessage: null,
    },
  ];

  export default DEFAULTEQNS