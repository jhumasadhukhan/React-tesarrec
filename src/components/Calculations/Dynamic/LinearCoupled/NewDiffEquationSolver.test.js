import { configure, shallow, render } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import DefaultEqns from "../../../../containers/Lab/DefaultStates/DefaultEqns";
import DefaultVars from "../../../../containers/Lab/DefaultStates/DefaultVars";
import NewDiffEquationSolver from "./NewDiffEquationSolver";
import props from "../SampleEquations/linear5";

// configure({ adapter: new Adapter() });
// describe("<NewDiffEquationSolver/>", () => {
//   it("should solve on differential eqn ", () => {
//     let myEqns = [];

//     DefaultEqns.forEach((eqn) => {
//       myEqns.push(eqn.TextEqn);
//     });

//     let myVars = {};

//     DefaultVars.forEach((VarElement) => {
//       myVars[VarElement.LatexForm] = VarElement.VarCurrent;
//     });

//     const firstNow = performance.now();
//     let computedResults = NewDiffEquationSolver({
//       method:"RK4",
//       h: 0.1,
//       t0: 0.5,
//       numberOfCycles: 10,
//       eqns: myEqns,
//       vars: myVars, // { K_1=0.27}
//       LineNames: ["a", "b", "c", "d"],
//       initialConditions: [0.5, 0.5, 0.5, 0.5], // Y1, Y2
      
//     });
//     const secondNow = performance.now();
//     const howLongDidOurLoopTake = secondNow - firstNow;
//     console.log(howLongDidOurLoopTake);

//     //expect(wrapper.props().numberOfCycles).to.equal([ 'Euler', 'Midpoint', 'Runge Kutta' ]);

//     //console.log(wrapper.find('EulerData').debug({ verbose: true }));
//     //console.log(wrapper.debug({ verbose: true }));

//     expect(computedResults).toMatchSnapshot();
//   });
// });

describe("<NewDiffEquationSolver3rdPartyCheck/>", () => {
  it("should solve on differential eqn ", () => {
    let eqns = props.eqns
    let NUMCYCLES = props.NUMCYCLES
    let varObjects = props.varObjects
    let actualSolutionArr = props.actualSolutionArr
    let h = props.h
    let t0 = props.t0
    let initialConditions = props.initialConditions
   
    let calcedArr = 
      NewDiffEquationSolver({
        method:"RK4",
        h:h,
        numberOfCycles:NUMCYCLES,
        eqns:eqns,
        vars:varObjects, // { K_1=0.27}
        LineNames:["a","b"],
        initialConditions:initialConditions ,// y1
        t0:t0
      });
      console.log(calcedArr)
    expect(calcedArr).toMatchSnapshot();
  });
});


