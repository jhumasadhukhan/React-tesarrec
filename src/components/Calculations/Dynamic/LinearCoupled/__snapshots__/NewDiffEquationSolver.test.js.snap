// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewDiffEquationSolver3rdPartyCheck/> should solve on differential eqn  1`] = `
Array [
  Array [
    1,
    0,
  ],
  Array [
    1.0050000000000001,
    0.1,
  ],
  Array [
    1.02,
    0.2,
  ],
  Array [
    1.045,
    0.30000000000000004,
  ],
  Array [
    1.0799999999999998,
    0.4,
  ],
  Array [
    1.1249999999999996,
    0.5,
  ],
  Array [
    1.1799999999999995,
    0.6,
  ],
  Array [
    1.2449999999999997,
    0.7,
  ],
  Array [
    1.3199999999999996,
    0.7999999999999999,
  ],
  Array [
    1.4049999999999996,
    0.8999999999999999,
  ],
  Array [
    1.4999999999999996,
    0.9999999999999999,
  ],
]
`;
